File: programs/merkle-distributor/src/instructions/clawback.rs
// Instruction to clawback funds once they have expired

use anchor_lang::{context::Context, prelude::*, Accounts, Key, Result};
use anchor_spl::{
    token,
    token::{Token, TokenAccount},
};

use crate::{error::ErrorCode, state::merkle_distributor::MerkleDistributor};

/// [merkle_distributor::clawback] accounts.
#[derive(Accounts)]
pub struct Clawback<'info> {
    /// The [MerkleDistributor].
    #[account(mut)]
    pub distributor: Account<'info, MerkleDistributor>,

    /// Distributor ATA containing the tokens to distribute.
    #[account(
        mut,
        associated_token::mint = distributor.mint,
        associated_token::authority = distributor.key(),
        address = distributor.token_vault
    )]
    pub from: Account<'info, TokenAccount>,

    /// The Clawback token account.
    #[account(mut)]
    pub to: Account<'info, TokenAccount>,

    /// Admin account
    /// Only admin can claw back
    #[account(mut, address = distributor.admin @ ErrorCode::Unauthorized)]
    pub admin: Signer<'info>,

    /// The [System] program.
    pub system_program: Program<'info, System>,

    /// SPL [Token] program.
    pub token_program: Program<'info, Token>,
}

/// Claws back unclaimed tokens by:
/// 1. Transferring remaining funds from the vault to the clawback receiver
/// 2. Marking the distributor as clawed back
#[allow(clippy::result_large_err)]
pub fn handle_clawback(ctx: Context<Clawback>) -> Result<()> {
    let distributor = &ctx.accounts.distributor;

    require!(!distributor.clawed_back, ErrorCode::ClawbackAlreadyClaimed);

    let seeds = [
        b"MerkleDistributor".as_ref(),
        &distributor.mint.to_bytes(),
        &distributor.creator.to_bytes(),
        &distributor.version.to_le_bytes(),
        &[ctx.accounts.distributor.bump],
    ];

    token::transfer(
        CpiContext::new(
            ctx.accounts.token_program.to_account_info(),
            token::Transfer {
                from: ctx.accounts.from.to_account_info(),
                to: ctx.accounts.to.to_account_info(),
                authority: ctx.accounts.distributor.to_account_info(),
            },
        )
        .with_signer(&[&seeds[..]]),
        ctx.accounts.from.amount,
    )?;

    let distributor = &mut ctx.accounts.distributor;

    distributor.clawed_back = true;

    Ok(())
}

File: bindings/ts/raw_instructions/clawback.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface ClawbackAccounts {
  /** The [MerkleDistributor]. */
  distributor: PublicKey
  /** Distributor ATA containing the tokens to distribute. */
  from: PublicKey
  /** The Clawback token account. */
  to: PublicKey
  /**
   * Admin account
   * Only admin can claw back
   */
  admin: PublicKey
  /** The [System] program. */
  systemProgram: PublicKey
  /** SPL [Token] program. */
  tokenProgram: PublicKey
}

export function clawback(
  accounts: ClawbackAccounts,
  programId: PublicKey = PROGRAM_ID
) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.distributor, isSigner: false, isWritable: true },
    { pubkey: accounts.from, isSigner: false, isWritable: true },
    { pubkey: accounts.to, isSigner: false, isWritable: true },
    { pubkey: accounts.admin, isSigner: true, isWritable: true },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
    { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([111, 92, 142, 79, 33, 234, 82, 27])
  const data = identifier
  const ix = new TransactionInstruction({ keys, programId, data })
  return ix
}
